enhancements/machine-config/mco-network-configuration.md

```diff
@@ -0,0 +1,253 @@
+
+## Motivation
+
+With hybrid clouds, node-networking setup is becoming even more challenging.
+Different payloads have different networking requirements, and not everything
+can be satisfied as overlays on top of the main interface of the node (e.g.
+SR-IOV, L2, other L2).
+
+The [Container Network Interface](https://github.com/containernetworking/cni) (CNI) standard enables different solutions for connecting networks on the node with pods. Some of them are [part of the standard](https://github.com/containernetworking/plugins), and there are others that extend support for [Open vSwitch bridges](https://github.com/kubevirt/ovs-cni), [SR-IOV](https://github.com/hustcat/sriov-cni), and more...
+
+However, in all of these cases, the node must have the networks setup before the pod is scheduled. Setting up the networks in a dynamic and heterogenous cluster, with dynamic networking requirements, is a challenge by itself - and this is what this project is addressing.
+
+The node network configuration shall remain in the MachineConfigOperator workflow, but be declarative.
+
+### Goals
+
+* Enable an admin to modify node networking configuration without rebooting (Day 2)
+* Enable cluster administrators to modify node networking after cluster installation (day 2).
+* Integration with MachineConfig node configuration for consistency, familiarity and roll out functionality.
+* Rebootless configuration without deferring errors to a later reboot
+* Rolling back cluster breaking configuration mistakes.
+* 
+### Non-Goals
+
+* Replace SRIOV operator
+
+## Proposal
+
+#### Option A
+
+The node networking configuration shall driven by MCO (and by extension kubernetes-nmstate) and executed by
+[nmstate](https://nmstate.github.io/).
+
+Do a KubeletConfig like integration:
+
+* NodeNetworkState will be introduced. This allows the user to see the network state of a node in the same declarative language that can be used to configure it.
+* The user will configure with NodeNetworkConfigurationPolicy (NNCP). This is a declaritive method of defining network state, to be applied by nmstate.
+* NNCP, just like KubeletConfig, targets a MachinePool
+* The MachineConfigDaemon (**MCD**) vendors the knmstate configuring/rollback functionality.
+* MCD runs a vendored subprocess that keeps NodeNetworkState (**NNS**) up to date with the OpenShift API
+* NMState is invoked with Podman (Python project with dependencies outside of stdlib)
+
+1. User Creates **NNCP**
+2. KNMStateController takes the **NNCP** desiredConfig and puts it as a **MachineConfig** (MC) for the desired pool
+3. MachineConfigController renders a new config and its update subcontroller starts updating desiredState of the Nodes
+4. MachineConfigDaemon sees the diff between currentState and desiredState
+5. In the case all the new rendered MC is equal except the **networkState** section, instead of reboot 
```

Comment:
Do we plan to support concurrent NNCP requests?

Reply:
What would constitute a concurrent request?

Reply:
The admin should create the NNCP CRs, right?  
Does below use case is valid ?

1. Admin creates NNCP#1 CR
2. nmstate & MCO processing NNCP#1 CR request (it could take some time)
3. Admin creates another NNCP#2 CR (while nmstate&MCO still processing the request from step 2).

Reply:
It depends on how MCO would handle that scenario. I don't have that knowledge yet.